### PR TITLE
Updates plugins to use the Catalog and Bindings interfaces

### PR DESCRIPTION
**Pull request by RCHowell**
## Relevant Issues

#1496 

## Description

This updates the our plugins to be based on the catalog interface defined in partiql-planner. It flips the dependency from planner->spi to be spi->planner.

## Other Information
- Updated Unreleased Section in CHANGELOG: **No**
- Any backward-incompatible changes? **YES**
- Any new external dependencies? **NO**
- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **YES**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/partiql-memory/src/main/kotlin/org/partiql/plugins/memory/MemoryConnector.kt [hunk 1]
```diff
@@ -16,45 +16,79 @@
...
+    /**
+     * Implement [Catalog] over the tables map.
+     */
+    private val catalog = object : Catalog {
+
+        override fun getName(): String = name
+
+        override fun getTable(name: Name): Table? {
+            if (name.hasNamespace()) {
+                error("MemoryCatalog does not support namespaces")
```
Comment (alancai98) [inline plugins/partiql-memory/src/main/kotlin/org/partiql/plugins/memory/MemoryConnector.kt:85]: Is `MemoryCatalog` something that will be added in a subsequent PR? Or should this refer to MemoryConnector's catalog:

```suggestion
                error("MemoryConnector's Catalog does not support namespaces")
```

File: plugins/partiql-memory/README.md [hunk 2]
```diff
@@ -1,36 +1,19 @@
...
+val catalog = MemoryCatalog.builder()
+    .name("hello")
+    .defineTable("pi", MemoryTable(
+        type = PType.typeFloat32(),
+        data = ionFloat(3.14),
+    ))
+    .build()
+
+// create a connector to be used in a session
+val connector = MemoryConnector.from(catalog)
```
Comment (alancai98) [inline plugins/partiql-memory/README.md:18]: nit: code snippet slightly out of date. guess we can update after all the APIs are more finalized

- MemoryCatalog -> MemoryConnector
- MemoryTable(...) -> MemoryTable.of()
- data should be of type `Datum`
- MemoryConnector.from() does not exist

**Review by alancai98 — APPROVED** [review]
plugin update following spi changes looks good. some minor comments related to `LocalConnector`

File: plugins/partiql-memory/src/main/kotlin/org/partiql/plugins/memory/MemoryConnector.kt [hunk 3]
```diff
@@ -16,45 +16,79 @@
...
+    internal class Factory : Connector.Factory {
 
         override val name: String = "memory"
 
-        override fun create(catalogName: String, config: StructElement?): MemoryConnector {
-            val catalog = catalogs.firstOrNull { it.name == catalogName }
-                ?: error("Catalog $catalogName is not registered in the MemoryPlugin")
-            return MemoryConnector(catalog)
+        override fun create(config: StructElement): Connector {
+            TODO("Instantiation of a MemoryConnector via the factory is currently not supported")
```
[the remark this answers is not in the document]
Comment (RCHowell) [reply plugins/partiql-memory/src/main/kotlin/org/partiql/plugins/memory/MemoryConnector.kt:46]: Yes. It's required for conformance testing.

**Pull request merged**